### ISSUE TITLE
should ignore all merge commits when bumping version

### ISSUE THREAD

**Issue by bochen2014** [lerna/lerna#1377]
here is npm version info
```
 ─┬ lerna@2.10.1   
    └─┬ conventional-changelog-cli@1.3.21   
         └── conventional-changelog@1.1.23    
```
I'm using `lerna` 's  independent version strategy, and here is my commit tree
```
*   commit#4  -- this is a merge commit; should this be ignored when bumping?
|\
|   * commit#3  chore(release): lerna publish  [ci skip] ***NO_CI*** (tag: package1@1.0.1)
*   |  commit#2 chore(package2): update package2 only
|/
* commit#1 chore(package1): update package1 only
* base (tag: package1@1.0.0 , package2@1.0.0 )
```

the scenario is as below:
A developer committed a commit#1 on `package1`, and pushed to origin. This triggered a build, once the build is finished ,lerna committed commit#3 and also bumpt `package1` 's version to `1.0.1`;
while the build was in progress, another developer committed another change commit#2 on `package2`. The developer then merged in commit#3 and this result in commit#4.

Now when developer push commit#4, because commit#4 contains changes in commit#3 (which is `package.json` and `CHANGELOG.md` of `package1`),  `package1` gets bumped again. 

so `package1` will end up with `1.0.2` which doesn't make sense.

Is there a way to let lerna ignore commit#3 (and **all merge commits* ) while it calculate version? thanks

**Comment by evocateur** [comment]
This is one reason you shouldn't publish from a feature branch.

Ignoring merge commits and whatnot is actually the domain of `conventional-changelog`, not lerna (assuming you're using `lerna publish --conventional-commits`, you didn't include any config or other context).

Again, the simplest solution here is to prohibit non-canary publish from feature branches with `--allow-branch master` (or `command.publish.allowBranch = "master"` in lerna.json)

**Comment by bochen2014** [comment]
thanks for your reply. 

No I am not publishing from feature branch.  everything happens in master branch here. It's purely a timing issue -- while CI is building the project , another developer already commit a new change. 
when CI finishes, lerna will commit as well. 
So the 2nd developer will have to merge in the lerna's commit, which will cause the same package to bump again (as lerna updates pacakge.json and CHNAGELOG.md)

does that make sense?

**Comment by evocateur** [comment]
Oh, yeah, that makes more sense. One of the drawbacks of non-canary releases in CI, at the moment. I would like to fix this, but it’s a dramatic change that I haven’t had time to fully-specify. Something akin to semantic-release, perhaps even reusing SR plugins or providing our own lifecycle plugin.

**Comment by bochen2014** [comment]
using `canary` release will only have impact on the final version number, it wont' change the fact of unnecessary bump. i.e. the change that CI build server commits only has changes in 'package.json' and 'CHANGELOG.md' and therefore should be ignored in the next round of change detection;

the real challenge is how can we spot this kind of changes and filter them out from change list during the next change detection; one possible way is by checking commit message (my CI build server  has formatted commit messages like 'chore(release): publish ***NO_CI***' so we can take advantage of this but again -- the change needs to be made at either *lerna* level or *conventional-version-bump* level

I reckon it's more likely that the change needs to be made in *conventional-version-bump*

**Comment by evocateur** [comment]
> I reckon it's more likely that the change needs to be made in conventional-version-bump

Definitely what I was thinking. There might be a better filter for its commit parser, for sure.

**Comment by soederpop** [comment]
I think you could fix this by not using the `--first-parent` argument to `git describe` which I think is in `@lerna/describe-ref`

I'm currently experiencing a problem where `lerna updated` uses tag 1.4.4 to compare against, when the latest release was 1.5.9.  This is due to the presence of merge commits from github PRs.

this is what lerna runs to determine the tag to compare against and see what's changed

```shell
$ git describe --first-parent --always --long --dirty # hahaha wtf git?
# v1.4.4-19-g20693b56
```

if instead it runs the following, it finds the correct tag.

```shell
$ git describe --tags --always --long --dirty
# v1.5.9-20-g20693b56
```
this ignores the merge commits since the `--first-parent` flag isn't being set.

i'm sure this `--first-parent` behavior is useful for something, so maybe my team is doing something wrong in our github workflow? For now I'm considering patching lerna to work around it.

**Comment by Feiyang1** [comment]
I'd like to have an option to ignore all merge commits in lerna changed command. 
We have a release branch we use as a staging area for the next release while the development continues in the master branch. Eventually we will publish the release candidate in the release branch and merge it back to the master branch. 
In the next release cycle, the updates in the merge commit are seen as changes and the same package gets bumped again even though it doesn't have any change except for the version bump in package.json from the last release.

Is there any plan to address this use case?

**Comment by evocateur** [comment]
@Feiyang1 Please create a separate issue. Commenting on a closed issue isn't (normally) going to get you anywhere.
